File: qc-application/src/main/java/com/eager/questioncloud/application/post/PostService.java
package com.eager.questioncloud.application.post;

import com.eager.questioncloud.common.PagingInformation;
import com.eager.questioncloud.domain.post.Post;
import com.eager.questioncloud.domain.post.PostContent;
import com.eager.questioncloud.domain.post.PostDetail;
import com.eager.questioncloud.domain.post.PostListItem;
import com.eager.questioncloud.domain.post.PostRepository;
import com.eager.questioncloud.exception.CustomException;
import com.eager.questioncloud.exception.Error;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class PostService {
    private final PostPermissionChecker postPermissionChecker;
    private final PostRepository postRepository;

    public Post register(Post post) {
        if (!postPermissionChecker.hasPermission(post.getWriterId(), post.getQuestionId())) {
            throw new CustomException(Error.FORBIDDEN);
        }

        return postRepository.save(post);
    }

    public List<PostListItem> getPostList(Long userId, Long questionId, PagingInformation pagingInformation) {
        if (!postPermissionChecker.hasPermission(userId, questionId)) {
            throw new CustomException(Error.FORBIDDEN);
        }

        return postRepository.getPostList(questionId, pagingInformation);
    }

    public int countPost(Long questionId) {
        return postRepository.count(questionId);
    }

    public PostDetail getPostDetail(Long userId, Long postId) {
        if (!postPermissionChecker.hasPermission(userId, postId)) {
            throw new CustomException(Error.FORBIDDEN);
        }
        return postRepository.getPostDetail(postId);
    }

    public void modify(Long postId, Long userId, PostContent postContent) {
        Post post = postRepository.findByIdAndWriterId(postId, userId);
        post.updateQuestionBoardContent(postContent);
        postRepository.save(post);
    }

    public void delete(Long postId, Long userId) {
        Post post = postRepository.findByIdAndWriterId(postId, userId);
        postRepository.delete(post);
    }
}

File: qc-domain/qc-post/qc-post-api/src/main/kotlin/com/eager/questioncloud/post/service/PostService.kt
package com.eager.questioncloud.post.service

import com.eager.questioncloud.common.exception.CoreException
import com.eager.questioncloud.common.exception.Error
import com.eager.questioncloud.common.pagination.PagingInformation
import com.eager.questioncloud.post.domain.Post
import com.eager.questioncloud.post.domain.PostContent
import com.eager.questioncloud.post.dto.PostDetail
import com.eager.questioncloud.post.dto.PostPreview
import com.eager.questioncloud.post.implement.PostPermissionChecker
import com.eager.questioncloud.post.implement.PostReader
import com.eager.questioncloud.post.infrastructure.repository.PostRepository
import org.springframework.stereotype.Service

@Service
class PostService(
    private val postPermissionChecker: PostPermissionChecker,
    private val postReader: PostReader,
    private val postRepository: PostRepository,
) {
    fun register(post: Post): Post {
        if (!postPermissionChecker.hasPermission(post.writerId, post.questionId)) {
            throw CoreException(Error.FORBIDDEN)
        }

        return postRepository.save(post)
    }

    fun getPostPreviews(userId: Long, questionId: Long, pagingInformation: PagingInformation): List<PostPreview> {
        return postReader.getPostPreviews(userId, questionId, pagingInformation)
    }

    fun countPost(questionId: Long): Int {
        return postReader.countPost(questionId)
    }

    fun getPostDetail(userId: Long, postId: Long): PostDetail {
        return postReader.getPostDetail(userId, postId)
    }

    fun modify(postId: Long, userId: Long, postContent: PostContent) {
        val post = postRepository.findByIdAndWriterId(postId, userId)
        post.updateQuestionBoardContent(postContent)
        postRepository.save(post)
    }

    fun delete(postId: Long, userId: Long) {
        val post = postRepository.findByIdAndWriterId(postId, userId)
        postRepository.delete(post)
    }
}
